modules/face/src/facemarkLBF.cpp

```diff
@@ -0,0 +1,1294 @@
+
+        faceDetector(image.getMat(), faces, extra_params);
+
+        return true;
+    }
+
+    bool FacemarkLBFImpl::configFaceDetector(){
+        if(!isSetDetector){
+            /*check the cascade classifier file*/
+            std::ifstream infile;
+            infile.open(params.cascade_face.c_str(), std::ios::in);
+            if (!infile) {
+               std::string error_message = "The cascade classifier model is not found.";
+               CV_Error(CV_StsBadArg, error_message);
+
+               return false;
+            }
+
+            face_cascade.load(params.cascade_face.c_str());
+        }
+        return true;
+    }
+
+    bool FacemarkLBFImpl::defaultFaceDetector(const Mat image, std::vector<Rect> & faces){
+        Mat gray;
+
+        faces.clear();
+
+        if(image.channels()>1){
+            cvtColor(image,gray,CV_BGR2GRAY);
+        }else{
+            gray = image;
+        }
+
+        equalizeHist( gray, gray );
+        face_cascade.detectMultiScale( gray, faces, 1.05, 2, 0|CV_HAAR_SCALE_IMAGE, Size(30, 30) );
+
+        return true;
+    }
+
+    bool FacemarkLBFImpl::getData(void * items){
+        if(items!=0){
+            // do nothing
+        }
+        return true;
+    }
+
+    bool FacemarkLBFImpl::addTrainingSample(InputArray image, InputArray landmarks){
+        std::vector<Point2f> & _landmarks = *(std::vector<Point2f>*)landmarks.getObj();
+        configFaceDetector();
+        prepareTrainingData(image.getMat(), _landmarks, data_faces, data_shapes, data_boxes);
+        return true;
+    }
+
+    void FacemarkLBFImpl::training(void* parameters){
+        if(parameters!=0){/*do nothing*/}
+        if (data_faces.size()<1) {
+           std::string error_message =
+            "Training data is not provided. Consider to add using addTrainingSample() function!";
+           CV_Error(CV_StsBadArg, error_message);
+        }
+
+        // flip the image and swap the landmark position
+        data_augmentation(data_faces, data_shapes, data_boxes);
+
+        Mat mean_shape = getMeanShape(data_shapes, data_boxes);
+
+        int N = (int)data_faces.size();
+        int L = N*params.initShape_n;
+        std::vector<Mat> imgs(L), gt_shapes(L), current_shapes(L);
+        std::vector<BBox> bboxes(L);
+        RNG rng(getTickCount());
+        for (int i = 0; i < N; i++) {
+            for (int j = 0; j < params.initShape_n; j++) {
+                int idx = i*params.initShape_n + j;
+                int k = 0;
+                do {
+                    k = rng.uniform(0, N);
+                } while (k == i);
+                imgs[idx] = data_faces[i];
+                gt_shapes[idx] = data_shapes[i];
+                bboxes[idx] = data_boxes[i];
+                current_shapes[idx] = data_boxes[i].reproject(data_boxes[k].project(data_shapes[k]));
+            }
+        }
+
+        // random shuffle
+        unsigned int seed = (unsigned int)std::time(0);
+        std::srand(seed);
```

Comment:
`rand()` is not thread-safe - can't provide reproducible results (for testing).
Consider to use cv::rand() , cv::theRNG() instead.

Reply:
according to the documentation http://docs.opencv.org/3.2.0/d2/de8/group__core__array.html#ga75843061d150ad6564b5447e38e57722 theRNG function will use different random generator for each thread, but in this case the program only needs to randomize the order of training sample, which should produce same random order for each sample category (i.e. the randomized order or image sample should corresponds to the randomized order of the annotation). If reproducible results is expected, then i think we just need to set the seed value to a fixed number, or just remove this shuffling part (i don't think it will produce significantly different results)

Reply:
`cv::theRNG` is thread-safe (maintains own state for each thread).
Seed initialization from `time()` should be removed from algorithm implementations - seed should be controlled externally.

Regarding this exact algorithm, I believe it should work without shuffling of input data.